### ISSUE TITLE
mingw-w64: Most `debuginfo-gdb` tests fail with new GDB 7.11.1

### ISSUE THREAD

**Issue by petrochenkov** [rust-lang/rust#40184]
If you recently updated mingw version to `x86_64-6.3.0-win32-seh-rt_v5-rev1` or just updated packages in MSYS2 you may discover this unpleasant surprise:
```
running 1 test
test [debuginfo-gdb] debuginfo-gdb\union-smoke.rs ... FAILED

failures:

---- [debuginfo-gdb] debuginfo-gdb\union-smoke.rs stdout ----
        NOTE: compiletest thinks it is using GDB without native rust support
NOTE: compiletest thinks it is using GDB version 7011001

error: line not found in debugger output: $1 = {a = {__0 = 2 '\002', __1 = 2 '\002'}, b = 514}
status: exit code: 0
command: PATH=blahblahblah C:\mingw-w64\x86_64-6.3.0-win32-seh-rt_v5-rev1\mingw64\bin\gdb -quiet -batch -nx -command=C:\msys64\home\we\rust\build\build\x86_64-pc-windows-gnu\test\debuginfo\union-smoke.debugger.script
stdout:
------------------------------------------
GNU gdb (GDB) 7.11.1
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-w64-mingw32".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word".
Breakpoint 1 at 0x40156e: file C:/msys64/home/we/rust\src/test\debuginfo/union-smoke.rs, line 50.
[New Thread 6068.0x12f4]
quit
A debugging session is active.

        Inferior 1 [process 6068] will be killed.

Quit anyway? (y or n) [answered Y; input not from terminal]

------------------------------------------
stderr:
------------------------------------------
warning: SHIMVIEW: ShimInfo(Complete)
Failed to resume program execution (ContinueDebugEvent failed, error 87)

------------------------------------------
```

or even worse (with MSYS subsystem gdb)
```
running 1 test
test [debuginfo-gdb] debuginfo-gdb\union-smoke.rs ... FAILED

failures:

---- [debuginfo-gdb] debuginfo-gdb\union-smoke.rs stdout ----
        NOTE: compiletest thinks it is using GDB without native rust support
NOTE: compiletest thinks it is using GDB version 7011001

error: line not found in debugger output: $1 = {a = {__0 = 2 '\002', __1 = 2 '\002'}, b = 514}
status: exit code: 0
command: PATH=blahblahblah C:\msys64\usr\bin\gdb -quiet -batch -nx -command=C:\msys64\home\we\rust\build\build\x86_64-pc-windows-gnu\test\debuginfo\union-smoke.debugger.script
stdout:
------------------------------------------
GNU gdb (GDB) 7.11.1
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-pc-msys".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word".
Breakpoint 1 at 0x40156e: file C:/msys64/home/we/rust\src/test\debuginfo/union-smoke.rs, line 50.
[New Thread 6600.0x19d0]

------------------------------------------
stderr:
------------------------------------------
Warning: /home/we/rust/build/C: No such file or directory.
Warning: /msys64/home/we/rust./src/etc: No such file or directory.
warning: SHIMVIEW: ShimInfo(Complete)
C:\msys64\home\we\rust\build\build\x86_64-pc-windows-gnu\test\debuginfo\union-smoke.debugger.script:9: Error in sourced command file:
No symbol "u" in current context.
Failed to resume program execution (ContinueDebugEvent failed, error 87)

------------------------------------------
```

**Comment by petrochenkov** [comment]
Workaround:
- Purge MSYS's gdb package (`pacman -R gdb`). I'm not sure if its version can be rolled back given that MSYS2 uses Arch Linux's rolling release scheme.
- Install mingw-w64 with previous version of GDB. `x86_64-6.3.0-win32-seh-rt_v5-rev1` -> `x86_64-6.2.0-win32-seh-rt_v5-rev1`

**Comment by alexcrichton** [comment]
It looks like if the 6.3.0 gdb is run manually it produces the expected output, it's just when called by the build system that it fails as indicated.

**Comment by alexcrichton** [comment]
This is specifically caused when the stdin to the gdb process is a pipe, not inherited. I have absolutely no clue why. I've got a patch to fix though that'll go in #40777

**Comment by arielb1** [comment]
There's a cygwin mailing list thread that looks related: https://cygwin.com/ml/cygwin/2016-10/msg00235.html.

**Comment by arielb1** [comment]
From the mailing list:
> I've done some investigation, and the problem seems to be that when run from emacs or ddd, the inferior process doesn't end up in it's own process group, but I haven't made any progress on why that is happening. 

Anyone with a working `pc-windows-gnu` environment: can this be fixed by passing `CREATE_NEW_CONSOLE` or `CREATE_NEW_PROCESS_GROUP`?

**Comment by petrochenkov** [comment]
`x86_64-7.1.0-posix-seh-rt_v5-rev0` toolchain is still affected.

**Comment by petrochenkov** [comment]
Fixed starting with `x86_64-6.4.0-posix-seh-rt_v5-rev0` and `x86_64-7.1.0-posix-seh-rt_v5-rev2`!
